app/code/Magento/InventoryShipping/Observer/VirtualSourceDeductionProcessor.php

```diff
@@ -0,0 +1,271 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventoryShipping\Observer;
+
+use Magento\Catalog\Model\Product\Type as VirtualType;
+use Magento\Downloadable\Model\Product\Type as DownloadableType;
+use Magento\Framework\Event\ObserverInterface;
+use Magento\Framework\Event\Observer as EventObserver;
+use Magento\Framework\Exception\CouldNotSaveException;
+use Magento\Framework\Exception\InputException;
+use Magento\Framework\Validation\ValidationException;
+use Magento\InventoryReservations\Model\ReservationBuilderInterface;
+use Magento\InventoryReservationsApi\Api\AppendReservationsInterface;
+use Magento\InventoryApi\Api\SourceItemsSaveInterface;
+use Magento\InventorySales\Model\StockByWebsiteIdResolver;
+use Magento\Framework\Exception\LocalizedException;
+use Magento\InventoryShipping\Model\GetSourceItemBySourceCodeAndSku;
+use Magento\InventoryCatalog\Model\GetSkusByProductIdsInterface;
+use Magento\InventoryConfigurationApi\Api\GetStockItemConfigurationInterface;
+use Magento\InventorySourceSelection\Model\GetDefaultSourceSelectionAlgorithmCodeInterface;
+use Magento\InventorySourceSelectionApi\Api\Data\InventoryRequestInterface;
+use Magento\InventorySourceSelectionApi\Api\Data\InventoryRequestInterfaceFactory;
+use Magento\InventorySourceSelectionApi\Api\Data\ItemRequestInterfaceFactory;
+use Magento\InventorySourceSelectionApi\Api\Data\SourceSelectionResultInterface;
+use Magento\InventorySourceSelectionApi\Api\SourceSelectionServiceInterface;
+use Magento\Sales\Api\Data\InvoiceItemInterface;
+use Magento\Sales\Model\Order\Item;
+use Magento\Framework\Serialize\Serializer\Json;
+use Magento\InventoryCatalog\Api\DefaultSourceProviderInterface;
+
+/**
+ * Class VirtualSourceDeductionProcessor | Probably need to divide on services
+ */
+class VirtualSourceDeductionProcessor implements ObserverInterface
```

Comment:
`class VirtualSourceDeductionProcessor has a coupling between objects value of 23. Consider to reduce the number of dependencies under 13.`

23 external dependencies is big enough

Reply:
done